### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive when casting to HTMLElement (from Element)

### ISSUE THREAD

**Issue by bschlenk** [typescript-eslint/typescript-eslint#9007]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=FAYw9gdgzgLgBAUwDZwLxwCZhAVwLYIQwB0AjjggE4CeAysgiDGJQBQDkGAlgIZJgBzdgEo4PKHAASAFQCyAGQCiSBASLBgAek1wYErhABmVCQAMQACy5IMp3WDgAjBHDCHd1AA4vTy1YRg7AHcrSzguCQgweBgLFygeAjEzGQU-NUDQSFg4S2sMNDhWZDIKGnoVJhYOAG0eHGZDbBwoAF0ROAAfTsQkUXEpOSUVDKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const el = document.querySelector('dialog') as HTMLElement

// ts infers `child` to be of type `Element` which is not the same as `HTMLElement`
const child = (el.querySelector('[autofocus]') || el) as HTMLElement

// `focus` doesn't exist on Element, which is why we cast in the first place
child.focus()
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect this not to trigger, since `HTMLElement` is not the same as the `Element` type ts would infer.

### Actual Result

This rule triggers and automatically removes the type assertion, causing the following lines to start failing.

**Comment by kirkwaiblinger** [comment]
@bschlenk 
Interesting! This appears to have to do with overload resolution complications
```ts
interface AType {
    aProperty: string;
}

interface AChildType extends AType {
    aNotherProperty: string;
}

// no bug
declare const subType: AChildType
var needsToBeSubType = subType as AChildType;
needsToBeSubType.aNotherProperty;

// bug! 
declare function getTypeOrSubType<T extends AType>(): T;
// generic argument is inferred as AChildType with the assertion, AType without. 
var needsToBeSubtype = getTypeOrSubType() as AChildType;
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
needsToBeSubType.aNotherProperty;
```

**Comment by kirkwaiblinger** [comment]
There is a chance this could be pretty hard to solve, or might run into some known blockers, along the lines of https://github.com/typescript-eslint/typescript-eslint/issues/9007. Curious if anyone in @typescript-eslint/triage-team knows off the top of their head

**Comment by kirkwaiblinger** [comment]
fyi @bschlenk if you want a way around this bug, I might recommend the following as a workaround/better code anyway:
```ts
// just specify the type argument and no type assertion is needed in the first place
const el = document.querySelector<HTMLElement>('dialog');
const child = (el.querySelector<HTMLElement>('[autofocus]') || el);
child.focus();
```
https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=FAYw9gdgzgLgBAUwDZwLxwCZhAVwLYIQwB0AjjggE4CeAysgiDGJQBQDkGAlgIZJgBzdgEo4PKHAASAFQCyAGQCiSBASLBQkWHBAALLkgxo4rZGQo16KpiwA8MhctWEYAPg4BtHjmYAzbDhQALoicAA%2BYYhIwkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by bradzacher** [comment]
This is working as intended.
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/8388, https://github.com/typescript-eslint/typescript-eslint/issues/7207,
https://github.com/typescript-eslint/typescript-eslint/issues/5808, https://github.com/typescript-eslint/typescript-eslint/issues/5359, https://github.com/typescript-eslint/typescript-eslint/issues/528 etc

**Comment by kirkwaiblinger** [comment]
@bradzacher Do you think we should try to document this more prominently on the rule page? Seems to have high volume of duplicates for what's basically an unfixable bug

**Comment by bradzacher** [comment]
We could do. It doesn't come up too often but no harm in documenting it.
